### ISSUE TITLE
Tuple equality comparison limitations - Is this "problem" due to System.ValueTuple or VB?

### ISSUE THREAD

**Issue by craigajohnson** [dotnet/vblang#105]
From the code below:

```
        Dim z As (a As Integer, b As Integer) = (5, 10)
        Dim q = (6, 12)

        Dim ok_true = z.Equals((5, 10))
        Dim ok_false = q.Equals(z)

        Dim nocompile_true = (z = (5, 10))
        Dim nocompile_false = (q = z)
```

ok_true and ok_false work fine.

nocompile_true gives error:

`BC30452 Operator '=' is not defined for types '(a As Integer, b As Integer)' and '(Integer, Integer)'`

nocompile_false gives error:

`BC30452	Operator '=' is not defined for types '(Integer, Integer)' and '(a As Integer, b As Integer)'`

And comparing two literal tuples gives the same error:

`Dim nope = ((6, 10) = (6, 10))`

`BC30452 Operator '=' is not defined for types '(Integer, Integer)' and '(Integer, Integer)'`

So, is there a (good) reason that tuples can't be tested for equality using the = operator as long as all of the types of the individual elements are equatable? Besides the elements themselves, there isn't really anything else to compare, is there?

Could the compiler re-write the expression to perform individual element comparisons?

**Comment by AnthonyDGreen** [comment]
It's not a problem with ValueTuple and it's a [deliberate design decision for VB](https://github.com/dotnet/vblang/blob/master/meetings/2016/LDM-2016-05-06-VB.md).

We originally decided not to distribute any operators on tuples over their elements. I believe we've reconsidered in recent C# LDMs but I don't think we shipped it in 15.3. @jcouv and @jaredpar for comment.

**Comment by jcouv** [comment]
Yes, this is being reconsidered for C#, but the decision is not finalized. All I can say is this feature is definitely not part of C# 7.1 or 7.2. Here's the proposal to track the C# change: https://github.com/dotnet/csharplang/issues/190

@AnthonyDGreen If we decide to allow those operators on tuples in C#, I assume we would make the same change to VB. In other words, do we need two LDM decisions, or just one?

**Comment by craigajohnson** [comment]
I just read the link from the C# proposal above in its entirety and was amused.

Given that discussion, I see no inherent confusion/violation of principle in having the compiler simply apply the = operator to each element in turn, thus saving unnecessary manual effort. Special handling in the compilers is already occurring for the System.ValueTuple type so it seems justifiable/defensible to follow that fruitful path a little farther to accommodate this proposal.

**Comment by AnthonyDGreen** [comment]
@jcouv we'd probably go the same way but we can make those decisions independently, yes.

@craigajohnson I suppose it's simple enough to say two tuple instances are equatable if each of their elements are equatable. If not there's a compile-time error. This would use the user defined `=` and `<>` operators, not .Equals.

**Comment by Nukepayload2** [comment]
If VB supports extension operators, this problem will not exist.

```VB.NET
Module ValueTupleExtensions

    <ExtensionOperator>
    Public Operator =(tuple1 As (Integer, Integer), tuple2 As (Integer, Integer)) As Boolean
        Return tuple1.Item1 = tuple2.Item1 AndAlso tuple1.Item2 = tuple2.Item2
    End Operator

    <ExtensionOperator>
    Public Operator <>(tuple1 As (Integer, Integer), tuple2 As (Integer, Integer)) As Boolean
        Return tuple1.Item1 <> tuple2.Item1 OrElse tuple1.Item2 <> tuple2.Item2
    End Operator

End Module
```
